[VLM] Feat: new mm scheduler that supports batch computation of mm embedding across multiple requests and mm_items assignment

yangsijia-celina opened the pull request:
<!-- Thank you for your contribution! Please follow these guidelines to enhance your pull request. If anything is unclear, submit your PR and reach out to maintainers for assistance. Join our Slack community at https://slack.sglang.io to discuss further. -->

## Motivation

<!-- Describe the purpose and goals of this pull request. -->
implement multimodal scheduling logic that batches mm_items from different requests for embedding computation, as described in https://github.com/sgl-project/sglang/issues/14331. By batching mm embeddings across multiple requests, we can better utilize GPU resources and achieve improved performance—especially in small-image scenarios, where the gains are more pronounced.

## Modifications

<!-- Detail the changes made in this pull request. -->

1. Add a server argument `enable_batch_compute_mm_embeddings` to toggle this feature on/off.
2. In `manager/mm_utils`, collect uncached mm_items across all requests in current scheduling batch and compute their embeddings together, instead of processing per request.
3. Add `scheduler_multimodal.py` and introduce `MMDPScheduler` to handle:
     a. mm_item assignment across DP instances based on `MMDPSchedulePolicy`. Currently only `MMDPSchedulePolicy.LOAD_BALANCE` is supported. The placement logic matches the existing `run_dp_sharded_mrope_vision_model` behavior; effectively, we move the placement from the model layer up to the scheduler layer.
     b. How mm_items are grouped on a single DP instance according to `MMPackPolicy`. Currently only `ALL_PACK` is implemented, i.e., all mm_items assigned to this instance are computed together.
     c. Distributing pixel_values, grid_thws, etc., to each DP instance and gathering the embedding results.
4. On the model side (currently only adapted for `qwen3_vl`), when embedding batch compute is enabled, we skip the original `run_dp_sharded_mrope_vision_model` path since DP assignment has already been moved to the scheduler.

Note: This `batch_compute_mm_embeddings` capability is not coupled to DP. Embedding batch computation can be used even without DP enabled.

**Benefits**

1. Better performance, especially for small-image workloads. See the benchmark results below for details.
2. Improved extensibility of multimodal scheduling:
    a. By moving multimodal scheduling logic upper to scheduler layer, we can apply a unified scheduling logic across different models without heavy per-model changes, keeping model-layer complexity low. Optimizations from [#14907](https://github.com/sgl-project/sglang/pull/14907) and [#14977](https://github.com/sgl-project/sglang/pull/14977) can also be integrated here.
    b. We can introduce richer `MMDPSchedulePolicy` and `MMPackPolicy` options later to achieve optimal performance across different scenarios.

**TODO**

- [ ] Refine this feature based on review comments.
- [ ] Merge with the main branch (note: the current main(until 3619ec6 commit) seems to have some multimodal accuracy issues).
- [ ] Add video support to MMDPScheduler.
- [ ] Support more models.

Feedback welcome!! Looking forward to your comments, reviews, and discussion!

## Accuracy Tests

<!-- If this pull request affects model outputs (e.g., changes to the kernel or model forward code), provide accuracy test results. -->
v0.5.6.post2
<img width="1538" height="300" alt="image" src="https://github.com/user-attachments/assets/7438fe68-ecf7-4102-aab7-5052a7627ceb" />

This PR
<img width="1362" height="294" alt="image" src="https://github.com/user-attachments/assets/9d54b087-dfcd-4771-bc6c-a712d0e60537" />


## Benchmarking and Profiling

<!-- If this pull request impacts inference speed, provide benchmarking and profiling results. -->
Compared with v0.5.6.post2
num_prompt: 128, qps=con=32, 224*224, 1 image, 64/1, dp=1

without feature
```
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    32.0      
Max request concurrency:                 32        
Successful requests:                     128       
Benchmark duration (s):                  3.71      
Total input tokens:                      13311     
Total input text tokens:                 4863      
Total input vision tokens:               8448      
Total generated tokens:                  128       
Total generated tokens (retokenized):    128       
Request throughput (req/s):              34.53     
Input token throughput (tok/s):          3590.65   
Output token throughput (tok/s):         34.53     
Peak output token throughput (tok/s):    40.00     
Peak concurrent requests:                58        
Total token throughput (tok/s):          3625.18   
Concurrency:                             16.15     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   467.81    
Median E2E Latency (ms):                 444.42    
---------------Time to First Token----------------
Mean TTFT (ms):                          467.79    
Median TTFT (ms):                        444.41    
P99 TTFT (ms):                           791.10    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

with feature(enable_batch_compute_mm_embeddings)
```
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    32.0      
Max request concurrency:                 32        
Successful requests:                     128       
Benchmark duration (s):                  3.43      
Total input tokens:                      13310     
Total input text tokens:                 4862      
Total input vision tokens:               8448      
Total generated tokens:                  128       
Total generated tokens (retokenized):    128       
Request throughput (req/s):              37.27     
Input token throughput (tok/s):          3875.04   
Output token throughput (tok/s):         37.27     
Peak output token throughput (tok/s):    42.00     
Peak concurrent requests:                44        
Total token throughput (tok/s):          3912.31   
Concurrency:                             5.23      
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   140.33    
Median E2E Latency (ms):                 141.95    
---------------Time to First Token----------------
Mean TTFT (ms):                          140.31    
Median TTFT (ms):                        141.94    
P99 TTFT (ms):                           215.83    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

num_prompt: 128, qps=con=32, 224*224, 5 image, 64/1, dp=2

without feature
```
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    32.0      
Max request concurrency:                 32        
Successful requests:                     128       
Benchmark duration (s):                  6.74      
Total input tokens:                      47136     
Total input text tokens:                 4896      
Total input vision tokens:               42240     
Total generated tokens:                  128       
Total generated tokens (retokenized):    128       
Request throughput (req/s):              19.00     
Input token throughput (tok/s):          6997.74   
Output token throughput (tok/s):         19.00     
Peak output token throughput (tok/s):    34.00     
Peak concurrent requests:                59        
Total token throughput (tok/s):          7016.75   
Concurrency:                             26.42     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1390.45   
Median E2E Latency (ms):                 1564.65   
---------------Time to First Token----------------
Mean TTFT (ms):                          1390.43   
Median TTFT (ms):                        1564.64   
P99 TTFT (ms):                           1744.77   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```

with feature(enable_batch_compute_mm_embeddings)
```
============ Serving Benchmark Result ============
Backend:                                 sglang-oai-chat
Traffic request rate:                    32.0      
Max request concurrency:                 32        
Successful requests:                     128       
Benchmark duration (s):                  5.32      
Total input tokens:                      47154     
Total input text tokens:                 4914      
Total input vision tokens:               42240     
Total generated tokens:                  128       
Total generated tokens (retokenized):    128       
Request throughput (req/s):              24.08     
Input token throughput (tok/s):          8869.85   
Output token throughput (tok/s):         24.08     
Peak output token throughput (tok/s):    32.00     
Peak concurrent requests:                64        
Total token throughput (tok/s):          8893.92   
Concurrency:                             25.01     
----------------End-to-End Latency----------------
Mean E2E Latency (ms):                   1038.59   
Median E2E Latency (ms):                 1097.38   
---------------Time to First Token----------------
Mean TTFT (ms):                          1038.57   
Median TTFT (ms):                        1097.37   
P99 TTFT (ms):                           1371.44   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          0.00      
Median TPOT (ms):                        0.00      
P99 TPOT (ms):                           0.00      
---------------Inter-Token Latency----------------
Mean ITL (ms):                           0.00      
Median ITL (ms):                         0.00      
P95 ITL (ms):                            0.00      
P99 ITL (ms):                            0.00      
Max ITL (ms):                            0.00      
==================================================
```
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
3. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
4. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) (`/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`) or contact authorized users to do so.
5. After green CI and required approvals, ask Merge Oncalls to merge.

yangsijia-celina committed:
feat: support vit batch compute across multiple reqs

Signed-off-by: yangsijia.614 <yangsijia.614@bytedance.com>

yangsijia-celina committed:
feat: support vit batch compute across multiple reqs with dp

Signed-off-by: yangsijia.614 <yangsijia.614@bytedance.com>

yangsijia-celina committed:
refactor

Signed-off-by: yangsijia.614 <yangsijia.614@bytedance.com>

yangsijia-celina committed:
code clean

Signed-off-by: yangsijia.614 <yangsijia.614@bytedance.com>

kousakawang commented on python/sglang/srt/managers/mm_utils.py:
```diff
@@ -454,6 +475,201 @@ def _get_chunked_prefill_embedding(
+        # step2.1: get mm items to be computed on current rank according to MMDPSchedulePolicy and MMPackPolicy
+
+        # if enable encoder DP, first apply MMDPSchedulePolicy
+        if len(embedding_items) == 0:
```
fix len(embedding_items) == 0

kousakawang commented on python/sglang/srt/managers/mm_utils.py:
```diff
@@ -454,6 +475,201 @@ def _get_chunked_prefill_embedding(
+        # print(f"{grid_thw_list_local=}")
+
+        # apply MMPackPolicy
+        embedding_items_list_local_rank = MMDPScheduler.get_mm_pack_result(
```
maybe it's better to apply pack policy before concat all pixel_values and grid_thw, or we need to split and concat them again (for all_pack policy,  it's ok)

yangsijia-celina committed:
rebase

fix: merge with main and resolve some comments
fix: adapt to EmbeddingResult
feat: support vit schedule for qwen2.5-vl and fix some bugs
refactor MMPackPolicy
rm fixed num pack policy

Signed-off-by: yangsijia.614 <yangsijia.614@bytedance.com>

yangsijia-serena commented:
/rerun-failed-ci

luzengxiangcn commented:
Does this PR support Qwen3-vl-embedding?

Commit:
fix import

Commit:
mm-pack merge in process

Commit:
fix vit-pack bug

Commit:
remove time waste

Commit:
move to device

Commit:
remove fallback

kousakawang committed:
Merge branch 'main' into feat/vit-schedule
